sxt/multiexp/bucket_method/accumulation.h

```diff
@@ -0,0 +1,170 @@
+  exponents_viewable_data.reset();
+
+  // combine partial sums
+  memmg::managed_array<T> bucket_sums_dev{bucket_sums.size(), &resource};
+  combine_partial_bucket_sums<<<dim3(255, num_outputs, 1), num_bytes, 0, stream>>>(
+      bucket_sums_dev.data(), partial_bucket_sums.data(), num_blocks);
+  partial_bucket_sums.reset();
+  basdv::async_copy_to_device(bucket_sums, bucket_sums_dev, stream);
+  co_await xendv::await_stream(stream);
+}
+
+template <bascrv::element T>
+xena::future<> accumulate_buckets_impl(basct::span<T> bucket_sums, basct::cspan<T> generators,
+                                       basct::cspan<const uint8_t*> exponents,
+                                       size_t split_factor) noexcept {
+  constexpr size_t bucket_group_size = 255;
+  constexpr size_t num_bucket_groups = 32;
+  static constexpr unsigned num_bytes = 32; // hard code to 32 for now
+  auto num_outputs = exponents.size();
+  SXT_DEBUG_ASSERT(
+      // clang-format off
+      bucket_sums.size() == bucket_group_size * num_bucket_groups * num_outputs &&
+      (bucket_sums.empty() || basdv::is_active_device_pointer(bucket_sums.data()))
+      // clang-format on
+  );
+
+  // Pick some reasonable values for min and max chunk size so that
+  // we don't run out of GPU memory or split computations that are
+  // too small.
+  //
+  // Note: These haven't been informed by much benchmarking. I'm
+  // sure there are better values. This is just putting in some
+  // ballpark estimates to get started.
+  size_t min_chunk_size = 1ull << 10u;
+  size_t max_chunk_size = 1ull << 20u;
+  if (num_outputs > 0) {
+    max_chunk_size = basn::divide_up(max_chunk_size, num_outputs);
+    min_chunk_size *= num_outputs;
+    min_chunk_size = std::min(max_chunk_size, min_chunk_size);
+  }
+
+  auto [first, last] = basit::split(basit::index_range{0, generators.size()}
+                                        .min_chunk_size(min_chunk_size)
+                                        .max_chunk_size(max_chunk_size),
+                                    split_factor);
+  auto num_chunks = std::distance(first, last);
+
+  basdv::stream stream;
+  memr::async_device_resource resource{stream};
+  memmg::managed_array<T> partial_bucket_sums_data{&resource};
+  basct::span<T> partial_bucket_sums;
+  if (num_chunks > 1) {
+    partial_bucket_sums_data.resize(num_bucket_groups * bucket_group_size * num_chunks);
+    partial_bucket_sums = partial_bucket_sums_data;
+  } else {
+    partial_bucket_sums = bucket_sums;
+  }
+
+  size_t step = bucket_group_size * num_bucket_groups * num_outputs;
+  size_t i = 0;
+  co_await xendv::concurrent_for_each(first, last, [&](const basit::index_range& rng) noexcept {
+    return accumulate_buckets_impl(partial_bucket_sums.subspan(step * i++, step), generators,
+                                   exponents, rng);
+  });
+  if (num_chunks <= 1) {
+    co_return;
+  }
+  combine_partial_bucket_sums<<<dim3(255, num_outputs, 1), num_bytes, 0, stream>>>(
```

Comment:
Does the `255` here relate to the `bucket_group_size`?

Reply:
Yes. I cleaned up the naming for magic numbers.